File: Parallel-Csharp/src/Parallel/Linq101/MyLinq/MatchMethods.cs
using Parallel.Linq101.DataSources;

namespace Parallel.Linq101.MyLinq;

public class MatchMethods {
    public List<Product> GetProductList() => Products.ProductList;

    public List<Customer> GetCustomerList() => Customers.CustomerList;

    //original:4 target:40 form Quantifiers
    public bool IntAllMatch() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        bool allMatch = numbers.All(n => n > 3);
        return allMatch;
    }

    public bool OddIntAllMatch() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        bool allMatch = numbers.Where(i => i % 2 == 1).All(n => n > 3);
        return allMatch;
    }

    public bool IntSquareAllMatch() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        bool allMatch = numbers.Select(i => i * i).All(n => n > 3);
        return allMatch;
    }

    public bool IntAllMatchInput(int i) {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        bool allMatch = numbers.All(n => n > i);
        return allMatch;
    }

    public bool IntAllMatchInput(int[] numbers, int i) {
        bool allMatch = numbers.All(n => n > i);
        return allMatch;
    }

    public bool StringAllMatch() {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        bool allMatch = words.All(n => string.CompareOrdinal(n, "b") > 0);
        return allMatch;
    }

    public bool LongStringAllMatch() {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        bool allMatch = words.Where(w => w.Length > 5).All(n => string.CompareOrdinal(n, "b") > 0);
        return allMatch;
    }

    public bool StringLengthAllMatch() {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        bool allMatch = words.Select(w => w.Length).All(n => n > 5);
        return allMatch;
    }

    public bool StringAllMatch(char c) {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        bool allMatch = words.All(w => w[0] > c);
        return allMatch;
    }

    public bool StringAllMatchCount(string[] strings, char c) {
        bool allMatch = strings.All(w => w[0] > c);
        return allMatch;
    }

    public bool ProductAllMatch() {
        List<Product> products = GetProductList();
        bool allMatch = products.All(p => p.UnitsInStock > 0);
        return allMatch;
    }

    public bool ExpensiveProductAllMatch() {
        List<Product> products = GetProductList();
        bool allMatch = products.Where(p => p.UnitPrice > 100).All(p => p.UnitsInStock > 0);
        return allMatch;
    }

    public bool CategoryProductAllMatch() {
        List<Product> products = GetProductList();
        bool allMatch = products.Where(p => p.Category == "Beverages").All(p => p.UnitsInStock > 0);
        return allMatch;
    }

    public Dictionary<string, bool> ProductEachGroupAllMatch() {
        List<Product> products = GetProductList();
        Dictionary<string, bool> allMatch = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key, g => g.All(p => p.UnitsInStock > 0));
        return allMatch;
    }

    public Dictionary<string, bool> ExpensiveProductEachGroupAllMatch() {
        List<Product> products = GetProductList();
        Dictionary<string, bool> allMatch = products.Where(p => p.UnitPrice > 100).GroupBy(p => p.Category)
            .ToDictionary(g => g.Key, g => g.All(p => p.UnitsInStock > 0));
        return allMatch;
    }

    public bool CustomerOrdersAllMatch() {
        List<Customer> customers = GetCustomerList();
        bool allMatch = customers.SelectMany(c => c.Orders)
            .All(o => o.OrderDate > new DateTime(2000, 1, 1));
        return allMatch;
    }

    public bool CustomerOrdersNestedAllMatch() {
        List<Customer> customers = GetCustomerList();
        bool allMatch = customers.SelectMany(c => c.Orders, (c, o) => Tuple.Create(
            c.CustomerId, o.OrderDate)).All(t => t.Item2 > new DateTime(2000, 1, 1));
        return allMatch;
    }

    public List<bool> CustomerOrdersEachNestedAllMatch() {
        List<Customer> customers = GetCustomerList();
        List<bool> allMatch = customers.Select(c => c.Orders.All(o =>
            o.OrderDate > new DateTime(2000, 1, 1))).ToList();
        return allMatch;
    }

    public bool WaCustomerOrdersNestedAllMatch() {
        List<Customer> customers = GetCustomerList();
        bool allMatch = customers.Where(c=>c.Region=="WA").SelectMany(c => c.Orders, (c, o) => Tuple.Create(
            c.CustomerId, o.OrderDate)).All(t => t.Item2 > new DateTime(2000, 1, 1));
        return allMatch;
    }

    public List<bool> WaCustomerOrdersEachNestedAllMatch() {
        List<Customer> customers = GetCustomerList();
        List<bool> allMatch = customers.Where(c=>c.Region=="WA").Select(c => c.Orders.All(o =>
            o.OrderDate > new DateTime(2000, 1, 1))).ToList();
        return allMatch;
    }
    
     public bool IntAnyMatch() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        bool anyMatch = numbers.Any(n => n > 3);
        return anyMatch;
    }

    public bool OddIntAnyMatch() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        bool anyMatch = numbers.Where(i => i % 2 == 1).Any(n => n > 3);
        return anyMatch;
    }

    public bool IntSquareAnyMatch() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        bool anyMatch = numbers.Select(i => i * i).Any(n => n > 3);
        return anyMatch;
    }

    public bool IntAnyMatchInput(int i) {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        bool anyMatch = numbers.Any(n => n > i);
        return anyMatch;
    }

    public bool IntAnyMatchInput(int[] numbers, int i) {
        bool anyMatch = numbers.Any(n => n > i);
        return anyMatch;
    }

    public bool StringAnyMatch() {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        bool anyMatch = words.Any(n => string.CompareOrdinal(n, "b") > 0);
        return anyMatch;
    }

    public bool LongStringAnyMatch() {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        bool anyMatch = words.Where(w => w.Length > 5).Any(n => string.CompareOrdinal(n, "b") > 0);
        return anyMatch;
    }

    public bool StringLengthAnyMatch() {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        bool anyMatch = words.Select(w => w.Length).Any(n => n > 5);
        return anyMatch;
    }

    public bool StringAnyMatch(char c) {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        bool anyMatch = words.Any(w => w[0] > c);
        return anyMatch;
    }

    public bool StringAnyMatchCount(string[] strings, char c) {
        bool anyMatch = strings.Any(w => w[0] > c);
        return anyMatch;
    }

    public bool ProductAnyMatch() {
        List<Product> products = GetProductList();
        bool anyMatch = products.Any(p => p.UnitsInStock > 0);
        return anyMatch;
    }

    public bool ExpensiveProductAnyMatch() {
        List<Product> products = GetProductList();
        bool anyMatch = products.Where(p => p.UnitPrice > 100).Any(p => p.UnitsInStock > 0);
        return anyMatch;
    }

    public bool CategoryProductAnyMatch() {
        List<Product> products = GetProductList();
        bool anyMatch = products.Where(p => p.Category == "Beverages").Any(p => p.UnitsInStock > 0);
        return anyMatch;
    }

    public Dictionary<string, bool> ProductEachGroupAnyMatch() {
        List<Product> products = GetProductList();
        Dictionary<string, bool> anyMatch = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key, g => g.Any(p => p.UnitsInStock > 0));
        return anyMatch;
    }

    public Dictionary<string, bool> ExpensiveProductEachGroupAnyMatch() {
        List<Product> products = GetProductList();
        Dictionary<string, bool> anyMatch = products.Where(p => p.UnitPrice > 100).GroupBy(p => p.Category)
            .ToDictionary(g => g.Key, g => g.Any(p => p.UnitsInStock > 0));
        return anyMatch;
    }

    public bool CustomerOrdersAnyMatch() {
        List<Customer> customers = GetCustomerList();
        bool anyMatch = customers.SelectMany(c => c.Orders)
            .Any(o => o.OrderDate > new DateTime(2000, 1, 1));
        return anyMatch;
    }

    public bool CustomerOrdersNestedAnyMatch() {
        List<Customer> customers = GetCustomerList();
        bool anyMatch = customers.SelectMany(c => c.Orders, (c, o) => Tuple.Create(
            c.CustomerId, o.OrderDate)).Any(t => t.Item2 > new DateTime(2000, 1, 1));
        return anyMatch;
    }

    public List<bool> CustomerOrdersEachNestedAnyMatch() {
        List<Customer> customers = GetCustomerList();
        List<bool> anyMatch = customers.Select(c => c.Orders.Any(o =>
            o.OrderDate > new DateTime(2000, 1, 1))).ToList();
        return anyMatch;
    }

    public bool WaCustomerOrdersNestedAnyMatch() {
        List<Customer> customers = GetCustomerList();
        bool anyMatch = customers.Where(c=>c.Region=="WA").SelectMany(c => c.Orders, (c, o) => Tuple.Create(
            c.CustomerId, o.OrderDate)).Any(t => t.Item2 > new DateTime(2000, 1, 1));
        return anyMatch;
    }

    public List<bool> WaCustomerOrdersEachNestedAnyMatch() {
        List<Customer> customers = GetCustomerList();
        List<bool> anyMatch = customers.Where(c=>c.Region=="WA").Select(c => c.Orders.Any(o =>
            o.OrderDate > new DateTime(2000, 1, 1))).ToList();
        return anyMatch;
    }
}

File: Parallel-Java/src/parallel/linq101/mylinq/MatchMethods.java
package parallel.linq101.mylinq;

import org.javatuples.Pair;
import parallel.linq101.datasource.Customer;
import parallel.linq101.datasource.Data;
import parallel.linq101.datasource.Product;

import java.time.LocalDate;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.*;

public class MatchMethods {
    public List<Product> getProductList() {
        return Data.getProductList();
    }

    public List<Customer> getCustomerList() {
        return Data.getCustomerList();
    }

    public boolean intAllMatch() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        boolean allMatch = Arrays.stream(numbers).allMatch(n -> n > 3);
        return allMatch;
    }

    public boolean oddIntAllMatch() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        boolean allMatch = Arrays.stream(numbers).filter(i -> i % 2 == 1).allMatch(n -> n > 3);
        return allMatch;
    }

    public boolean intSquareAllMatch() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        boolean allMatch = Arrays.stream(numbers).map(i -> i * i).allMatch(n -> n > 3);
        return allMatch;
    }

    public boolean intAllMatchInput(int i) {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        boolean allMatch = Arrays.stream(numbers).allMatch(n -> n > i);
        return allMatch;
    }

    public boolean intAllMatchInput(int[] numbers, int i) {
        boolean allMatch = Arrays.stream(numbers).allMatch(n -> n > i);
        return allMatch;
    }

    public boolean stringAllMatch() {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        boolean allMatch = Arrays.stream(words).allMatch(n -> n.compareTo("b") > 0);
        return allMatch;
    }

    public boolean longStringAllMatch() {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        boolean allMatch = Arrays.stream(words).filter(w -> w.length() > 5).allMatch(n -> n.compareTo("b") > 0);
        return allMatch;
    }

    public boolean stringLengthAllMatch() {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        boolean allMatch = Arrays.stream(words).map(w -> w.length()).allMatch(n -> n > 5);
        return allMatch;
    }

    public boolean stringAllMatch(char c) {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        boolean allMatch = Arrays.stream(words).allMatch(w -> w.charAt(0) > c);
        return allMatch;
    }

    public boolean stringAllMatchCount(String[] strings, char c) {
        boolean allMatch = Arrays.stream(strings).allMatch(w -> w.charAt(0) > c);
        return allMatch;
    }

    public boolean productAllMatch() {
        List<Product> products = getProductList();
        boolean allMatch = products.stream().allMatch(p -> p.getUnitsInStock() > 0);
        return allMatch;
    }

    public boolean expensiveProductAllMatch() {
        List<Product> products = getProductList();
        boolean allMatch = products.stream().filter(p -> p.getUnitPrice() > 100).allMatch(p -> p.getUnitsInStock() > 0);
        return allMatch;
    }

    public boolean categoryProductAllMatch() {
        List<Product> products = getProductList();
        boolean allMatch = products.stream().filter(p -> p.getCategory() == "Beverages").allMatch(p -> p.getUnitsInStock() > 0);
        return allMatch;
    }

    public Map<String, Boolean> productEachGroupAllMatch() {
        List<Product> products = getProductList();
        Map<String, Boolean> allMatch = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().allMatch(p -> p.getUnitsInStock() > 0)));
        return allMatch;
    }

    public Map<String, Boolean> expensiveProductEachGroupAllMatch() {
        List<Product> products = getProductList();
        Map<String, Boolean> allMatch = products.stream().filter(p -> p.getUnitPrice() > 100).collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().allMatch(p -> p.getUnitsInStock() > 0)));
        return allMatch;
    }

    public boolean customerOrdersAllMatch() {
        List<Customer> customers = getCustomerList();
        boolean allMatch = customers.stream().flatMap(c -> c.getOrders().stream())
                .allMatch(o -> o.getOrderDate().isAfter(LocalDate.of(2000, 1, 1)));
        return allMatch;
    }

    public boolean customerOrdersNestedAllMatch() {
        List<Customer> customers = getCustomerList();
        boolean allMatch = customers.stream().flatMap(c -> c.getOrders().stream()
                        .map(o -> Pair.with(c.getCustomerId(), o.getOrderDate())))
                .allMatch(t -> t.getValue1().isAfter(LocalDate.of(2000, 1, 1)));
        return allMatch;
    }

    public List<Boolean> customerOrdersEachNestedAllMatch() {
        List<Customer> customers = getCustomerList();
        List<Boolean> allMatch = customers.stream().map(c -> c.getOrders().stream().allMatch(o ->
                o.getOrderDate().isAfter(LocalDate.of(2000, 1, 1)))).toList();
        return allMatch;
    }

    public boolean waCustomerOrdersNestedAllMatch() {
        List<Customer> customers = getCustomerList();
        boolean allMatch = customers.stream().filter(c -> c.getRegion() == "WA").flatMap(c -> c.getOrders().stream()
                        .map(o -> Pair.with(c.getCustomerId(), o.getOrderDate())))
                .allMatch(t -> t.getValue1().isAfter(LocalDate.of(2000, 1, 1)));
        return allMatch;
    }

    public List<Boolean> waCustomerOrdersEachNestedAllMatch() {
        List<Customer> customers = getCustomerList();
        List<Boolean> allMatch = customers.stream().filter(c -> c.getRegion() == "WA").map(c -> c.getOrders().stream().allMatch(o ->
                o.getOrderDate().isAfter(LocalDate.of(2000, 1, 1)))).toList();
        return allMatch;
    }

    public boolean intAnyMatch() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        boolean anyMatch = Arrays.stream(numbers).anyMatch(n -> n > 3);
        return anyMatch;
    }

    public boolean oddIntAnyMatch() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        boolean anyMatch = Arrays.stream(numbers).filter(i -> i % 2 == 1).anyMatch(n -> n > 3);
        return anyMatch;
    }

    public boolean intSquareAnyMatch() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        boolean anyMatch = Arrays.stream(numbers).map(i -> i * i).anyMatch(n -> n > 3);
        return anyMatch;
    }

    public boolean intAnyMatchInput(int i) {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        boolean anyMatch = Arrays.stream(numbers).anyMatch(n -> n > i);
        return anyMatch;
    }

    public boolean intAnyMatchInput(int[] numbers, int i) {
        boolean anyMatch = Arrays.stream(numbers).anyMatch(n -> n > i);
        return anyMatch;
    }

    public boolean stringAnyMatch() {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        boolean anyMatch = Arrays.stream(words).anyMatch(n -> n.compareTo("b") > 0);
        return anyMatch;
    }

    public boolean longStringAnyMatch() {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        boolean anyMatch = Arrays.stream(words).filter(w -> w.length() > 5).anyMatch(n -> n.compareTo("b") > 0);
        return anyMatch;
    }

    public boolean stringLengthAnyMatch() {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        boolean anyMatch = Arrays.stream(words).map(w -> w.length()).anyMatch(n -> n > 5);
        return anyMatch;
    }

    public boolean stringAnyMatch(char c) {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        boolean anyMatch = Arrays.stream(words).anyMatch(w -> w.charAt(0) > c);
        return anyMatch;
    }

    public boolean stringAnyMatchCount(String[] strings, char c) {
        boolean anyMatch = Arrays.stream(strings).anyMatch(w -> w.charAt(0) > c);
        return anyMatch;
    }

    public boolean productAnyMatch() {
        List<Product> products = getProductList();
        boolean anyMatch = products.stream().anyMatch(p -> p.getUnitsInStock() > 0);
        return anyMatch;
    }

    public boolean expensiveProductAnyMatch() {
        List<Product> products = getProductList();
        boolean anyMatch = products.stream().filter(p -> p.getUnitPrice() > 100).anyMatch(p -> p.getUnitsInStock() > 0);
        return anyMatch;
    }

    public boolean categoryProductAnyMatch() {
        List<Product> products = getProductList();
        boolean anyMatch = products.stream().filter(p -> p.getCategory() == "Beverages").anyMatch(p -> p.getUnitsInStock() > 0);
        return anyMatch;
    }

    public Map<String, Boolean> productEachGroupAnyMatch() {
        List<Product> products = getProductList();
        Map<String, Boolean> anyMatch = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().anyMatch(p -> p.getUnitsInStock() > 0)));
        return anyMatch;
    }

    public Map<String, Boolean> expensiveProductEachGroupAnyMatch() {
        List<Product> products = getProductList();
        Map<String, Boolean> anyMatch = products.stream().filter(p -> p.getUnitPrice() > 100).collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().anyMatch(p -> p.getUnitsInStock() > 0)));
        return anyMatch;
    }

    public boolean customerOrdersAnyMatch() {
        List<Customer> customers = getCustomerList();
        boolean anyMatch = customers.stream().flatMap(c -> c.getOrders().stream())
                .anyMatch(o -> o.getOrderDate().isAfter(LocalDate.of(2000, 1, 1)));
        return anyMatch;
    }

    public boolean customerOrdersNestedAnyMatch() {
        List<Customer> customers = getCustomerList();
        boolean anyMatch = customers.stream().flatMap(c -> c.getOrders().stream()
                        .map(o -> Pair.with(c.getCustomerId(), o.getOrderDate())))
                .anyMatch(t -> t.getValue1().isAfter(LocalDate.of(2000, 1, 1)));
        return anyMatch;
    }

    public List<Boolean> customerOrdersEachNestedAnyMatch() {
        List<Customer> customers = getCustomerList();
        List<Boolean> anyMatch = customers.stream().map(c -> c.getOrders().stream().anyMatch(o ->
                o.getOrderDate().isAfter(LocalDate.of(2000, 1, 1)))).toList();
        return anyMatch;
    }

    public boolean waCustomerOrdersNestedAnyMatch() {
        List<Customer> customers = getCustomerList();
        boolean anyMatch = customers.stream().filter(c -> c.getRegion() == "WA").flatMap(c -> c.getOrders().stream()
                        .map(o -> Pair.with(c.getCustomerId(), o.getOrderDate())))
                .anyMatch(t -> t.getValue1().isAfter(LocalDate.of(2000, 1, 1)));
        return anyMatch;
    }

    public List<Boolean> waCustomerOrdersEachNestedAnyMatch() {
        List<Customer> customers = getCustomerList();
        List<Boolean> anyMatch = customers.stream().filter(c -> c.getRegion() == "WA").map(c -> c.getOrders().stream().anyMatch(o ->
                o.getOrderDate().isAfter(LocalDate.of(2000, 1, 1)))).toList();
        return anyMatch;
    }

}
